docs/prover/checking/sanity.md

```diff
@@ -2,9 +2,241 @@ Rule Sanity Checks
+   if all the `require` and `if` conditions are removed.
+
+   For example, the following rule would be flagged by the assert-vacuity check:
+   
+   ```cvl
+   rule tautology {
+     uint x; uint y;
+     require x != y;
+     ...
+     assert x < 2 || x >= 2,
+      "x must be smaller than 2 or greater than or equal to 2";
+   }
+   ```
+  
+  
+   Since every `uint` satisfies the assertion, the assertion is tautological,
+   which is likely to be an error in the specification.
+   
+   **Checking vacuity for invariants**
+   
+   For invariants, vacuity is checked by converting it into a rule that asserts the invariant expression without any require statements. Since a rule would check the assertion for all arbitrary starting states, if the rule passes, it means that the expression being asserted is a tautology. The invariant, which checks the expression for a smaller set of states, would also be a tautology.
+   
+   The `sanityCheck` contract below has two state variables `address root` and `uint a`. Due to the zero address checks in the `constructor` and the `changeRoot` function, the root address can never be zero. The `rootNonZero` invariant asserts that the root address is never zero. When we run this invariant with the `--rule_sanity` `advanced` or `basic` options, the prover creates a rule similar to the `rootNonZeroRule` below. This rule would fail since the tool could assume a starting state where the `root` is 0. This means that the invariant expression is not a tautology and the invariant passes. On the other hand the `aGE0` invariant, when run without the `--rule_sanity` option, will pass [view report](https://vaas-stg.certora.com/output/11775/871cf37193c75d27542b/?anonymousKey=dde443c4a806021716e863a454561a6ad1543d2e) but when we run it with the `--rule_sanity` `advanced` or `basic` options, the prover creates a rule similar to the `aGE0Rule` below. This rule passes, indicating that the invariant expression is a tautology. The verification report shows that the invariant failed vacuity check [view report](https://vaas-stg.certora.com/output/11775/4c4cb65f65c75f013c63/?anonymousKey=0b6a843857e6ead8e1bb1f11b984fb6e3e9fb6a8). 
+
+    ```solidity
+     contract sanityCheck{
+     address root;
+     uint a;
+     constructor(address _root){
+         require(_root != address(0));
+         root = _root;
+     }
+     function changeRoot (address root_) public{
+         require(root_ != address(0));
+         root = root_;
+     }
+     }
+     ```
+   
+   
+    ```cvl
+    
+      // Not a tautology
+      invariant rootNonZero()
+         root() != 0
+
+      rule rootNonZero(){
+         assert root() != 0;
+      }
+
+      // Tautology
+      invariant aGE0()
+         a() >= 0
+
+      rule aGE0Rule{
+         assert a() >= 0;
+      }
+
+    ```
+    
+    **Checking vacuity for rules**
+    
+    For rules, checking for tautology requires checking each assertion to see if 
```

Comment:
When working with rules, checking for tautology requires inspecting each assertion to see if 
    it’s meaningful. To do this, we employ few different checks depending
    on the syntax of the assertion expression.

Reply:
```suggestion
When working with rules, checking for tautology requires inspecting each assertion to see if
it’s meaningful. To do this, we employ a few different checks depending
on the syntax of the assertion expression.
```